csharp/src/Ice/Ice1BinaryConnection.cs

```diff
@@ -0,0 +1,496 @@
+        private int _nextStreamId;
+        private Task _receiveTask = Task.CompletedTask;
+        private Action<int> _receivedCallback;
+        private Task _sendTask = Task.CompletedTask;
+        private Action<int> _sentCallback;
+        private readonly bool _warn;
+        private readonly bool _warnUdp;
+
+        private static readonly List<ArraySegment<byte>> _closeConnectionFrame =
+            new List<ArraySegment<byte>> { Ice1Definitions.CloseConnectionFrame };
+
+        private static readonly List<ArraySegment<byte>> _validateConnectionFrame =
+            new List<ArraySegment<byte>> { Ice1Definitions.ValidateConnectionFrame };
+
+        public async ValueTask CloseAsync(Exception exception, CancellationToken cancel)
+        {
+            if (!(exception is ConnectionClosedByPeerException))
+            {
+                // Write and wait for the close connection frame to be written
+                try
+                {
+                    await SendFrameAsync(0, _closeConnectionFrame, cancel).ConfigureAwait(false);
+                }
+                catch
+                {
+                    // Ignore
+                }
+            }
+
+            // Notify the transport of the graceful connection closure.
+            try
+            {
+                await Transceiver.ClosingAsync(exception, cancel).ConfigureAwait(false);
+            }
+            catch
+            {
+                // Ignore
+            }
+
+            // Wait for the connection closure from the peer
+            try
+            {
+                await _receiveTask.WaitAsync(cancel).ConfigureAwait(false);
+            }
+            catch
+            {
+                // Ignore
+            }
+        }
+
+        public ValueTask DisposeAsync()
+        {
+            Transceiver.ThreadSafeClose();
+            Transceiver.Destroy();
+            return default;
+        }
+
+        public async ValueTask HeartbeatAsync(CancellationToken cancel) =>
+            await SendFrameAsync(0, _validateConnectionFrame, cancel).ConfigureAwait(false);
+
+        public async ValueTask InitializeAsync(
+            Action heartbeatCallback,
+            Action<int> sentCallback,
+            Action<int> receivedCallback,
+            CancellationToken cancel)
+        {
+            _heartbeatCallback = heartbeatCallback;
+            _sentCallback = sentCallback;
+            _receivedCallback = receivedCallback;
+
+            // Initialize the transport
+            await Transceiver.InitializeAsync(cancel).ConfigureAwait(false);
+
+            ArraySegment<byte> readBuffer = default;
+            if (!Endpoint.IsDatagram) // Datagram connections are always implicitly validated.
+            {
+                if (_incoming) // The server side has the active role for connection validation.
+                {
+                    int offset = 0;
+                    while (offset < _validateConnectionFrame.GetByteCount())
```

Comment:
We can avoid calling `GetByteCount` more than once
```suggestion
                    int byteCount = _validateConnectionFrame.GetByteCount();
                    while (offset < byteCount)
```